tensorstore/kvstore/s3/s3_credential_provider_test.cc

```diff
@@ -0,0 +1,187 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "tensorstore/kvstore/s3/s3_credential_provider.h"
+
+#include <fstream>
+#include <iostream>
+#include <string>
+#include <utility>
+#include <vector>
+
+
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+#include "absl/strings/escaping.h"
+#include "absl/strings/match.h"
+#include "absl/time/clock.h"
+#include "tensorstore/internal/env.h"
+#include "tensorstore/internal/http/curl_handle.h"
+#include "tensorstore/internal/http/curl_transport.h"
+#include "tensorstore/internal/http/http_request.h"
+#include "tensorstore/internal/http/http_response.h"
+#include "tensorstore/internal/path.h"
+#include "tensorstore/internal/test_util.h"
+#include "tensorstore/util/result.h"
+#include "tensorstore/util/status.h"
+#include "tensorstore/util/status_testutil.h"
+#include "tensorstore/util/str_cat.h"
+
+namespace {
+
+using ::tensorstore::StrCat;
+using ::tensorstore::internal::JoinPath;
+using ::tensorstore::internal::GetEnv;
+using ::tensorstore::internal::SetEnv;
+using ::tensorstore::internal::UnsetEnv;
+using ::tensorstore::internal_kvstore_s3::CredentialProvider;
+using ::tensorstore::internal_kvstore_s3::GetS3CredentialProvider;
+
+
+class TestData : public tensorstore::internal::ScopedTemporaryDirectory {
+ public:
+  std::string WriteCredentialsFile() {
+    auto p = JoinPath(path(), "aws_config");
+    std::ofstream ofs(p);
+    ofs <<
+        "discarded_value = 500\n"
+        "\n"
+        "[default]\n"
+        "aws_access_key_id =AKIAIOSFODNN7EXAMPLE\n"
+        "aws_secret_access_key = wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY\n"
+        "aws_session_token= abcdef1234567890 \n"
+        "\n"
+        "[alice]\n"
+        "aws_access_key_id = AKIAIOSFODNN6EXAMPLE\n"
+        "aws_secret_access_key = wJalrXUtnFEMI/K7MDENG/bPxRfiCZEXAMPLEKEY\n";
+    return p;
+  }
+};
+
+class S3CredentialProviderTest : public ::testing::Test {
+protected:
+ TestData test_data;
+ std::string credentials_filename;
+
+ // Environment variables to save and restore during setup and teardown
+ std::map<std::string, std::optional<std::string>> saved_vars{
+    {"AWS_SHARED_CREDENTIALS_FILE", std::nullopt},
+    {"AWS_ACCESS_KEY_ID", std::nullopt},
+    {"AWS_SECRET_ACCESS_KEY", std::nullopt},
+    {"AWS_SESSION_TOKEN", std::nullopt},
+    {"AWS_DEFAULT_PROFILE", std::nullopt},
+    {"AWS_PROFILE", std::nullopt}
+ };
+
+ void SetUp() override {
+    for(auto &pair: saved_vars) {
```

Comment:
Saving the env vars shouldn't be necessary, since they are process scoped and should not leak past test execution.